docs/corelang.md

```diff
@@ -0,0 +1,1025 @@
+`round`, `floor`, or `ceil`, e.g.
+`1.2 | floor, round, ceil --> 1 1 2`.
+
+::: Compatibility
+
+`jq` uses floats for any number,
+meaning that it does not distinguish integers from floats.
+Many operations in jaq, such as array indexing,
+check whether the passed numbers are indeed integer.
+The motivation behind this is to avoid
+rounding errors that may silently lead to wrong results.
+For example,
+`[0, 1, 2] | .[1] --> 1`, whereas
+`[0, 1, 2] | .[1.0000000000000001]` yields
+an error in jaq as opposed to `1` in `jq`.
+
+Furthermore, `jq` prints NaN as `null`; e.g. `nan | tojson` yields `null`.
+In contrast, jaq prints NaN as `NaN`; e.g. `nan | tojson --> "NaN"`.
+As a result, in jaq, any value can be round-tripped with `tojson | fromjson`; e.g.
+`nan | tojson | fromjson | isnan --> true`.
+In `jq`, `tojson | fromjson` does not round-trip, because of
+`NaN` and `Infinity`.
+
+:::
+
+### Text strings
+
+A text string is an array of bytes that can be constructed using the syntax `"..."`.
+Here, `...` may contain any UTF-8 characters
+in the range from `U+0020` to `U+10FFFF`, excluding `'"'` and `'\'`.
+For example,
+`"Hello 東京!" --> "Hello 東京!"`.
+
+Furthermore, `...` may contain escape sequences, which
+start with `'\'` and are followed by one of:
+
+- `b`, `f`, `n`, `r`, or `t`
+- `'"'` or `'\'`
+- `uHHHH`, where `HHHH` is a hexadecimal number
+- `(f)`, where `f` is a jq filter (*string interpolation*)
+
+Most characters below `U+0020` can only be produced via a Unicode escape sequence, i.e.
+`"NUL = \u0000" --> "NUL = \u0000"`.
+
+A string containing an interpolated filter, such as
+`"...\(f)..."`, is equivalent to
+`"..." + (f | tostring) + "..."`.
+For example:
+
+- `"Hello \("Alice", "Bob") 😀!\n" -->
+  "Hello Alice 😀!\n" "Hello Bob 😀!\n"`
+- `41 | "The successor of \(.) is \(.+1)." --> "The successor of 41 is 42."`.
+
+A string is *identifier-like* if it matches the regular expression
+`[a-zA-Z_][a-zA-Z_0-9]*`.
+
+Strings may be prefixed with `@x`, where `x` is an identifier; e.g. `@uri`.
+In such a case, the filters of interpolated strings are
+piped through `@x` instead of `tostring`, i.e.
+`@x "...\(f)..."` is equivalent to
+`@x "..." + (f | @x) + @x "..."`.
+When `"..."` does not contain any interpolated filter, then
+`@x "..."` is equivalent to `"..."`.
+For example,
+`"-[]?" | @uri "https://gedenkt.at/jaq/?q=\(.)" -->
+"https://gedenkt.at/jaq/?q=-%5B%5D%3F"`.
+
+### Byte strings
+
+A byte string is an array of bytes that is *not* interpreted as (UTF-8) text.
+It can be produced from a text string via the filter [`tobytes`](#tobytes), e.g.
+`"Hello, world! 🙂" | tobytes --> b"Hello, world! \xf0\x9f\x99\x82"`.
+Currently, there is no native syntax in jaq to produce a byte string directly.
+
+Byte strings differ from text strings in a few regards; in particular,
+they can be [indexed](#indexing) and [sliced](#slicing) in constant time.
+That makes byte strings interesting e.g. for parsing binary formats.
```

Comment:
Maybe some example to make it clear slice unit is bytes instead of code points?
```sh
$ jaq-n '"fäbler" | tobytes, . | .[1:] | ., length'
b"\xc3\xa4bler"
6
"äbler"
5
```